pkg/registry/consul.go

```diff
@@ -216,15 +218,23 @@ func (c *ConsulClient) Put(ctx context.Context, key string, value []byte) error
 		registryOperationTimeTaken.WithLabelValues(env, "Put").Observe(time.Now().Sub(startTime).Seconds())
 	}()
 
-	_, err := c.client.KV().Put(&api.KVPair{
-		Key:   key,
-		Value: value,
-	}, nil)
-	return err
+	// Using transaction because a simple put does not fetch the modifyindex
+	putTxn := api.TxnOps{&api.TxnOp{KV: &api.KVTxnOp{Verb: api.KVSet, Key: key, Value: value}}}
+	ok, resp, _, err := c.client.Txn().Txn(putTxn, nil)
+	if !ok {
+		if err != nil {
+			return "", err
+		}
+		// There must be exactly 1 error
+		err = fmt.Errorf(resp.Errors[0].What)
+		return "", err
+	}
+
+	return strconv.FormatUint(resp.Results[0].KV.ModifyIndex, 10), err
```

Comment:
add safe checks before accessing nested field to avoid panics.
`len(resp.Results)>0 && resp.Results[0].KV !=nil`